src/amoc_metrics.erl

```diff
@@ -1,122 +1,88 @@
 -module(amoc_metrics).
 
--export([start/0, init/2, user_size/0]).
--export([update_time/2, update_counter/2, update_counter/1, update_gauge/2]).
+-behaviour(prometheus_collector).
+
+-export([start/0, start_predefined_metrics/1, init/2]).
+-export([update_counter/1, update_counter/2, update_gauge/2, update_time/2]).
+-export([deregister_cleanup/1, collect_mf/2]).
 
 -include_lib("kernel/include/logger.hrl").
 
 -type simple_name() :: atom() | [atom()].
 -type name() :: simple_name() | {strict, simple_name()}.
--type type() :: counters | times | gauge.
+-type type() :: counters | gauge | times | histogram | summary.
 
 -export_type([name/0]).
 
--define(AMOC_DEFAULT_METRICS_REPORTER, exometer_report_graphite).
--define(AMOC_METRICS_REPORTING_INTERVAL, timer:seconds(10)).
+%% ===================================================================
+%% API
+%% ===================================================================
 
--spec start() -> any().
+-spec start() -> boolean().
 start() ->
-    maybe_add_reporter(),
-    subsribe_default_metrics(),
-    maybe_init_predefined_metrics().
-
--spec init(type(), name()) -> ok.
-init(Type, Name) ->
-    ExName = make_name(Type, Name),
-    ExType = exometer_metric_type(Type),
-    create_metric_and_maybe_subscribe(ExName, ExType).
+    maybe_add_exporter().
 
--spec user_size() -> [{size, non_neg_integer()}].
-user_size() ->
-    [{size, amoc_users_sup:count_no_of_users()}].
+-spec start_predefined_metrics(atom()) -> any().
+start_predefined_metrics(App) ->
+    Preconfigured = application:get_env(App, predefined_metrics, []),
+    [init(Type, Name) || {Type, Name} <- lists:flatten(Preconfigured)].
 
--spec update_time(name(), integer()) -> ok.
-update_time(Name, Value) ->
-    ExName = make_name(times, Name),
-    exometer:update(ExName, Value).
+-spec init(type(), name()) -> ok.
+init(counters, Name) ->
+    prometheus_counter:new([{name, Name}]);
+init(gauge, Name) ->
+    prometheus_gauge:new([{name, Name}]);
+init(times, Name) ->
+    prometheus_summary:new([{name, Name}]);
```

Comment:
Here I'd suggest using a histogram (with a good set of buckets) instead, because for times like TTD, we really need the quantiles, while `prometheus_summary` would only give us the number of events and the sum of the values. Btw, `prometheus_quantile_summary` would not help, because it is completely broken and unusable (this is why are not using it in MIM). See https://github.com/esl/MongooseIM/blob/master/src/instrument/mongoose_instrument_prometheus.erl#L53 for reference.